lib/RestructureCFG/InlineDispatcherSwitch.cpp

```diff
@@ -509,6 +509,65 @@ static void processNestedWeavedSwitches(SwitchNode *Switch) {
   }
 }
 
+static RecursiveCoroutine<bool> containsNonLocalCF(ASTNode *Node) {
+  switch (Node->getKind()) {
+  case ASTNode::NK_List: {
+    SequenceNode *Seq = llvm::cast<SequenceNode>(Node);
+
+    bool HasNonLocalCF = false;
+    for (ASTNode *N : Seq->nodes()) {
+      HasNonLocalCF = HasNonLocalCF or rc_recur containsNonLocalCF(N);
+    }
+
+    rc_return HasNonLocalCF;
+  } break;
+  case ASTNode::NK_Scs: {
+
+    // We do not want to move around a `ScsNode` for any reason
+    rc_return false;
+  } break;
+  case ASTNode::NK_If: {
+    IfNode *If = llvm::cast<IfNode>(Node);
+
+    bool HasNonLocalCF = false;
+    if (If->hasThen()) {
+      ASTNode *Then = If->getThen();
+      HasNonLocalCF = HasNonLocalCF or rc_recur containsNonLocalCF(Then);
+    }
+    if (If->hasElse()) {
+      ASTNode *Else = If->getElse();
+      HasNonLocalCF = HasNonLocalCF or rc_recur containsNonLocalCF(Else);
+    }
+
+    rc_return HasNonLocalCF;
+  } break;
+  case ASTNode::NK_Switch: {
+    auto *Switch = llvm::cast<SwitchNode>(Node);
+
+    bool HasNonLocalCF = false;
+    for (auto &LabelCasePair : Switch->cases()) {
+      HasNonLocalCF = HasNonLocalCF
+                      or rc_recur containsNonLocalCF(LabelCasePair.second);
+    }
+
+    rc_return HasNonLocalCF;
+  } break;
+  case ASTNode::NK_Set:
+  case ASTNode::NK_Code:
+  case ASTNode::NK_SwitchBreak: {
+    rc_return false;
+  } break;
```

Comment:
I'm not 100% sure about the `SwitchBreak`. Can you please elaborate?

Reply:
The reasoning is the following:
- A `SwitchBreak` can appear in the AST as the single element composing the body of a `case` of a `switch` node. Its meaning is equivalent to `nop` (e.g., a certain state variable assignment of an exit dispatcher corresponds to no statements executed).
- During the `inlineDispatcherSwitch` pass, in principle we can remove a `case` composed by a single `SwitchBreak` node. However, in the current state of things, this is done before in a dedicated pass called `simplifySwitchBreak`, indeed there is a specific point where we check that we do not encounter any `SwitchBreak` in this phase ([this check here](https://github.com/andrealinux1/revng-c/blob/954765bd27fbbbd5b28e738c30591490f8dc315e/lib/RestructureCFG/InlineDispatcherSwitch.cpp#L707).
- However, I share your concern that this is not crystal clear to understand.
- At the same time, I think that the semantics of the `containsNonLocalCF` helper function should handle the presence of a `SwitchBreak`. Maybe the non ideal thing here is the name of the helper itself. Basically we want an helper which replies "no you can't move around this `ASTNode`" only when it finds non local control flow related to loops (standard `break`s and `continue`s)

Reply:
I understand now, and I agree with your last statement.

Can you propose better naming for the helper?

Reply:
I changed the name of the helper [here](https://github.com/revng/revng-c/blob/6af6c8ffa81d1d570010e6e64e260a78eedd4ea9/lib/RestructureCFG/InlineDispatcherSwitch.cpp#L521), in `containsContinueOrBreak`